src/coreclr/vm/ilmarshalers.cpp

```diff
@@ -2496,10 +2496,14 @@ void ILBlittablePtrMarshaler::EmitConvertContentsNativeToCLR(ILCodeStream* pslIL
 
 bool ILBlittablePtrMarshaler::CanMarshalViaPinning()
 {
+    // [COMPAT] For correctness, we can't marshal via pinning if we might need to marshal differently at runtime.
+    // See calls to EmitExactTypeCheck where we check the runtime type of the object being marshalled.
```

Comment:
Just to understand the history: Was the `EmitExactTypeCheck` check added for better compatibility with .NET Framework, or was it added as attempt to fix a bug?

Reply:
It was added as an attempt to fix a compat issue from the StructMarshal IL stubs work. The old struct/class marshalling used the runtime type of a given type to determine how to marshal, not the metadata type. The Struct Marshalling IL Stubs initially used the metadata type. The type check was to bring back compat with the runtime type check because some customers were using it to map a native discriminated union to managed.